### ISSUE TITLE
consolidated metadata storage for Zarr v3

### ISSUE THREAD

**Issue by grlee77** [zarr-developers/zarr-specs#136]
In the current draft of the v3 spec in zarr-developers/zarr-python#898, consolidated metadata has been implemented similarly to v2. Currently this metadata gets stored in the group `meta/root/consolidated`, but this is not part of the spec. Likely we need to specify an extension for how consolidated metadata should be handled.

**Comment by ianliu** [comment]
Is the V3 considering metadata compression? I'm working with a dataset where the consolidated metadata (v2) has 1.2GB in size, and opening it from S3 takes a very long time to open.
After gzipping the metadata, it reduces to 20MB, so it would be nice to have compressed metadata.

**Comment by alimanfoo** [comment]
Hi @ianliu, thanks this is very useful to know. Adding a capability for compression into a v3 extension for consolidated metadata sounds like a good idea to me.

**Comment by rabernat** [comment]
@ianliu I'm curious what makes the metadata so big in your case? Do you have lots of arrays? Or very large custom attrs?

V3 allows for defining of custom [metadata encoding](https://zarr-specs.readthedocs.io/en/latest/core/v3.0.html#metadata-encoding), so a compressed format could be used.

Also, a major question for V3 is whether user metadata is a separate document or not (see https://github.com/zarr-developers/zarr-specs/pull/149#discussion_r942225058). If you have 1GB of custom attrs in a single array, that would be a strong motivation not to put user metadata together with the core array metadata.

**Comment by ianliu** [comment]
@rabernat in my particular case, I have many small arrays. It could be argued that there is a better data representation that would consolidate multiple arrays into a single array, thus reducing metadata, but the way our program is setup makes this a little bit trickier (mainly because our program produces results in parallel).

**Comment by joshmoore** [comment]
:100: for potential consolidated metadata features but

> but this is not part of the spec. Likely we need to specify an extension for how consolidated metadata should be handled.

would this suggest stripping the current consolidated metadata support in V3 until there's an extension?

**Comment by TomAugspurger** [comment]
I've drafted a version of this at https://github.com/TomAugspurger/zeps/blob/consolidated-metadata/draft/ZEP0006.md. A few questions before turning that into a PR:

Is this appropriate for a ZEP? Or would it go in the zarr-specs repo?

---

One major design decision is whether this consolidated metadata should be embedded in the `zarr.json`, or whether it should go in a file next to the root `zarr.json` (e.g. `consolidated_metadata.json`). Doing it in the `zarr.json` would make collecting all the metadata possible in a single HTTP request. With separate `zarr.json` and `consolidated_metadata.json` you would need two. The downside of putting it in `zarr.json` is potentially making *any* read of that file slow, even if you don't need the consolidated metadata.

I'm also a bit curious about how much async zarr-python reduces the pain of hierarchies with a large number of nodes. With dozens of HTTP requests in flight at the same time, maybe it's not so bad? If https://github.com/zarr-developers/zarr-specs/issues/284 (listing the child nodes in the root `zarr.json`) were implemented, then you'd be able to make all but the first HTTP request concurrently.

**Comment by jhamman** [comment]
I like the draft ZEP @TomAugspurger. Thanks for putting it together. 

As far as I understand, the current spec does not provide a way to add arbitrary extensions to a metadata document (this could be changed though). Provided we can address this challenge, we likely will want to support both an inline and external consolidated metadata object.

```json
// inline
"consolidated_metadata": {
    "kind": "inline":
    "metadata": {
        "air": {
          "shape": [2920, 25, 53],
          "fill_value": 0,
          ...
          "node_type": "array",
      },
      "lat": {
          "shape": [25],
          "fill_value": NaN,
          ...
          "node_type": "array",
      }
  }
}
// external
"consolidated_metadata": {
    "kind": "external":
    "metadata": "consolidated_metadata.json"
}
```

An alternative direction we could take is to define a Consolidated Metadata Store which could enable this functionality without needing to touch the root metadata document.

**Comment by TomAugspurger** [comment]
> the current spec does not provide a way to add arbitrary extensions to a metadata document

In the sense that that's just not spelled out anywhere as a thing that can be done? Or that adding additional keys might break things?

> we likely will want to support both an inline and external consolidated metadata object

I like the idea of future-proofing things by including a "kind" enum to indicate where to find the consolidated metadata, maybe we pick only one for the spec? This is minor, but I worry about putting too much complexity on the implementors.

**Comment by jhamman** [comment]
> In the sense that that's just not spelled out anywhere as a thing that can be done? Or that adding additional keys might break things?

I don't think we have defined the expected behavior for what will happen if an implementation encounters and unexpected key. So yes, I think thinks could break.

 > maybe we pick only one for the spec?

+1 one, pick one for now. We can always expand this later.

**Comment by TomAugspurger** [comment]
I'm not sure I see the motivation for disallowing extra keys in the metadata objects. That seems to make it hard to evolve the spec in forwards-compatible ways.

Anyway, maybe in this case it's fine to use because we're making a "future version of the spec"? And I guess we'll set `must_understand: false` to indicate that older versions shouldn't error, they should just ignore the extra key and fall back to the non-consolidated metadata.

**Comment by TomAugspurger** [comment]
Whoops, I just pushed this as a PR at https://github.com/zarr-developers/zarr-specs/pull/309, rather than a ZEP, sorry.

> That should definitely be some form of metadata in zarr.json. I'd very much be for using this as a way to kick the tires of the extension process. i.e., if you/we can't do it, then that needs to be fixed at the spec level.

I'll open a separate issue to discuss this. In my experience, STAC's method of each object containing a `stac_extensions` list with a list of extensions has worked well.
